src/factor.h

```diff
@@ -63,10 +71,18 @@ class CompactFactor {
   inline double potential(const VariableInFactor vifs[],
                           const variable_value_t var_values[],
                           const variable_id_t vid,
-                          const variable_value_t proposal) const {
+                          const variable_value_t proposal,
+                          feature_value_t val = DEFAULT_FEATURE_VALUE) const {
+    // For boolean, this is stored in CompactFactor.value;
+    //     caller shouldn't set the val arg.
+    // For categorical, this is stored in Factor.factor_params.value.
+    //     caller should pass in via the val arg.
+    // TODO: better data structures than the factor_params duct tape...
+    // See also: CompactFactorGraph::potential
+    val = (func_id == FUNC_AND_CATEGORICAL ? val : this->feature_value);
```

Comment:
Yes, we can do the lookup here to get rid of the `val` input. But then we could be calling `get_categorical_factor_params` 5 times instead of once (one in `update_weight`, and 4 calls to `potential` inside it)...

I think we just need to come up with a better way to store this lookup table...